### PR TITLE
Add binary crossentropy fallback functions

### PR DESCRIPTION
**Pull request by janEbert**
These fallbacks assume that if the first argument is a 1-element
`Vector` (as given by a model's output) and the second argument is a
scalar, the first argument should also be interpreted as a scalar.
Ref #850.

The use case is putting a model's scalar output right into the loss function instead of having to manually index into the 1-element `Vector` which is easily forgotten.

Not sure if this is desired, otherwise just close this. It does have the benefit of preventing not so obvious errors that – in my opinion – should not happen.

### CODE REVIEW

**Comment by MikeInnes** [comment]
Yeah, I'm not sure about this as is. Helpful coercions like this are tempting in each case, but adding them makes it gradually harder to tell what a function is actually meant to do. I think it might be preferable to document clearly that `binarycrossentropy` is meant to be used in a broadcasted way.

One other option is to make `binarycrossentropy` automatically broadcast if not given two scalars. This would actually address the same problem, but with a slightly more standard motivation (this lets us work with batches of outputs, which things like `crossentropy` already support via broadcasting.

**Comment by janEbert** [comment]
Yeah, too much magic is always bad. I can change it to your suggestion if you want that instead.

**Comment by CarloLucibello** [comment]
Not relevant anymore
